### ISSUE TITLE
TelemetryConfiguration during ConfigureServices

### ISSUE THREAD

**Issue by cijothomas** [Microsoft/ApplicationInsights-aspnetcore#631]
OriginalPosted by Nisheet Jain
I believe this is a bad pattern to follow?   
IServiceProvider serviceProvider = serviceCollection.BuildServiceProvider();    
https://github.com/Microsoft/ApplicationInsights-Kubernetes/blob/578f20e824e6248029554a1f8990b29c4a7c6d11/src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesModule.cs  

my issue is that I don’t have the TelemetryConfiguration available initially (during ConfigureServices) and I am looking for a lazy pattern to enable this.  Is there a workaround you can propose?

**Comment by nisheetjain** [comment]
@cijothomas , @SergeyKanzhelev any updates on this?

IMO we shouldn't try to create a partial ServiceProvider by calling serviceCollection.BuildServiceProvider();

**Comment by cijothomas** [comment]
@nisheetjain Please give me some info about what you are trying to achieve.
Are you trying to add a new TelemetryModule in ConfigureServices, and cannot do it since TelemetryConfiguration is not available?

**Comment by cijothomas** [comment]
You can add new TelemtryModule to the service collection like already done by SDK. https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/develop/src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs#L168

Before Configure() method is called, application insights sdk configures`TelemetryConfiguration` and calls Initialize(telemetryConfiguration) on all registered TelemetryModules.

**Comment by nisheetjain** [comment]
@cijothomas I get the correct instrumentation key in a lazy pattern.
this means that when I call 

> services.EnableKubernetes

I don't have the instrumentation key

Will EnableKubernetes work with out that?

**Comment by cijothomas** [comment]
Sorry I am not sure what is your specific question here. I am not so familiar with Kubernetes. :(
Can you rephrase your question to help me understand the issue you are trying to solve.

**Comment by SergeyKanzhelev** [comment]
@cijothomas they are evil - using `.Active` config. You may want to talk to them.

https://github.com/Microsoft/ApplicationInsights-Kubernetes/blob/578f20e824e6248029554a1f8990b29c4a7c6d11/src/ApplicationInsights.Kubernetes/Extensions/ApplicationInsightsExtensions.cs#L14-L26

@nisheetjain if you follow instructions and set env variable `APPINSIGHTS_INSTRUMENTATIONKEY` on container - this variable should be used as a key. So you do not need to specify it explicitly in code.

@cijothomas, btw, will this env variable logic work in case of multiple apps in the same process? Just curious.

**Comment by nisheetjain** [comment]
@SergeyKanzhelev I can't use the env variable since I don't have the key available at the deployment time unless you are saying I can set it at any point during application runtime and it will get picked

my scenario is something like
`services.EnableKubernetes((serviceProvider) => TelemetryConfiguration.Active.InstrumentationKey = {serviceProvider.GetService<IFoo>().GetInstrumentationKey});`

**Comment by SergeyKanzhelev** [comment]
Sure. So what's the question than?

+ @karolz-ms and @xiaomi7732 from ai-kubernetes repo

**Comment by xiaomi7732** [comment]
@nisheetjain If I understand right, you are trying to use application insights for Kubernetes and you are saying you won't have application insights key at the deployment time but you will have it in the runtime. That means, you can use it with in your code, is that correct?
1. EnableKubernetes() follows whatever the iKey the application insights SDK uses. That means, you can use whatever way is provided by the standard Application Insights SDK provide in this [document](
https://github.com/Microsoft/ApplicationInsights-aspnetcore/wiki/Getting-Started-with-Application-Insights-for-ASP.NET-Core#3-add-the-instrumentation-key).
2. If there is any further question specific to Application Insights for Kubernetes, feel free to open issue [here](https://github.com/Microsoft/ApplicationInsights-Kubernetes/issues).

Thanks,
Saar

**Comment by nisheetjain** [comment]
Thanks @xiaomi7732 that should work for me!!!

**Comment by cijothomas** [comment]
Closing as no further action on AI SDK side. Please reopen if there is something still unanswered
